Feature/delete default branch error

Ankit-Jaipuriar opened the pull request:
Closes #9224

## Change Description
### Background
When a user attempts to delete a repository's default branch, the API currently returns a generic HTTP transport error. This message is confusing as it doesn't explain why the action failed, leading users to believe there might be a network or server issue. This change improves the developer experience by providing a clear, actionable error message.

### Bug Fix
Problem - Deleting the default branch results in a cryptic error (...giving up after 5 attempt(s)) instead of a clear explanation.

Root cause - The API handler for branch deletion lacked a specific pre-check to see if the target branch was the default branch.

Solution - A check has been added to the DeleteBranch API handler. Before attempting the deletion, the handler now verifies if the branch is the default. If it is, the API returns a 403 Forbidden status with a user-friendly error message.

### New Feature
Not applicable. This is an enhancement to an existing error condition.

## Testing Details
The change was verified locally by compiling the lakefs server and lakectl client. After starting the server, I used lakectl to attempt to delete the default main branch and confirmed that the new, informative error message was correctly displayed in the terminal.

## Breaking Change?
No. This change only modifies an error response for a disallowed action and does not break any existing functionality for valid API or CLI calls.

## Additional info
CLI Output Before:
```
http:///api/v1/repositories//branches/main giving up after 5 attempt(s)
Error executing command.
```
CLI Output After:
```
Error executing command: you are trying to delete the default branch 'main', which is not supported.
```
## Contact Details
Provided via my GitHub account.

nopcoder reviewed (changes requested):
Graveler already handle the case for delete default branch by returning an error `ErrDeleteDefaultBranch`.

Suggest to just add a case in the controller.go to capture the error and return 403 (Forbidden) and the error value.

Ankit-Jaipuriar commented:
Hi @nopcoder,

Thanks for the helpful feedback! I've updated the pull request with all the requested changes. The code has been refactored to handle the existing error, and the test is updated to verify the 403 status code.

nopcoder commented on pkg/api/controller.go:
```diff
@@ -2898,8 +2898,11 @@ func (c *Controller) DeleteBranch(w http.ResponseWriter, r *http.Request, reposi
 	c.LogAction(ctx, "delete_branch", r, repository, branch, "")
-
 	err := c.Catalog.DeleteBranch(ctx, repository, branch, graveler.WithForce(swag.BoolValue(body.Force)))
+	if errors.Is(err, graveler.ErrDeleteDefaultBranch) {
+		writeError(w, r, http.StatusForbidden, err.Error())
+		return
+	}
```
can you move the check to `handleAPIErrorCallback` as another check under the `	case errors.Is(err, block.ErrForbidden)...` it will be cover cases when other code in the controller will call the delete branch in the future or test code.

nopcoder commented on pkg/api/controller_test.go:
```diff
@@ -2311,8 +2311,8 @@ func TestController_DeleteBranchHandler(t *testing.T) {
 		}
-		if resp.JSONDefault == nil {
-			t.Fatal("DeleteBranch expected error while trying to delete default branch")
+		if resp.StatusCode() != http.StatusForbidden {
```
Test `resp.JSON403 == nil` will check the status code and the content type as the generated code will populate this field just for this case.

Ankit-Jaipuriar commented:
Hi @nopcoder,

I've pushed the latest changes addressing all your feedback. The error handling is now in `handleAPIErrorCallback` and the test has been updated accordingly. Thanks!

Ankit-Jaipuriar commented on pkg/api/controller_test.go:
```diff
@@ -2311,8 +2311,8 @@ func TestController_DeleteBranchHandler(t *testing.T) {
-		if resp.JSONDefault == nil {
-			t.Fatal("DeleteBranch expected error while trying to delete default branch")
+		if resp.JSON403 == nil {
+			t.Fatal("DeleteBranch expected a 403 Forbidden response, but got none")
```
Hi @nopcoder, thanks for the approval! I saw your "nitpick" suggestion to report the actual status code. Should I go ahead and make that change, or are you planning to merge this as is once the checks pass? Let me know what you prefer. Thanks!
